build/utils/npm.go

```diff
@@ -9,189 +9,284 @@ import (
-		(*dependencies)[depId].Scopes = append((*dependencies)[depId].Scopes, scope)
+func parseNpmLsDependencyFunc(npmVersion *version.Version) func(data []byte) (*npmLsDependency, error) {
+	// If npm older than v7, use legacy struct for npm ls output.
+	if npmVersion.Compare("7.0.0") > 0 {
+		return legacyNpmLsDependencyParser
+	}
+	return npmLsDependencyParser
+}
+
+func legacyNpmLsDependencyParser(data []byte) (*npmLsDependency, error) {
+	legacyNpmLsDependency := new(legacyNpmLsDependency)
+	err := json.Unmarshal(data, &legacyNpmLsDependency)
+	if err != nil {
+		return nil, err
 	}
-	(*dependencies)[depId].RequestedBy = append((*dependencies)[depId].RequestedBy, pathToRoot)
+	return legacyNpmLsDependency.toNpmLsDependency(), nil
 }
 
-func scopeAlreadyExists(scope string, existingScopes []string) bool {
-	for _, existingScope := range existingScopes {
-		if existingScope == scope {
-			return true
+func npmLsDependencyParser(data []byte) (*npmLsDependency, error) {
+	npmLsDependency := new(npmLsDependency)
+	return npmLsDependency, json.Unmarshal(data, &npmLsDependency)
+}
+
+func appendDependency(dependencies map[string]*prototypeNode, dep *npmLsDependency, cacache *cacache, pathToRoot []string, log utils.Log) (err error) {
+	depId := dep.id()
+	scopes := dep.getScopes()
+	if dependencies[depId] == nil {
+		dependency := &prototypeNode{
+			Dependency:      entities.Dependency{Id: depId},
+			npmLsDependency: dep,
 		}
+
+		dependencies[depId] = dependency
+	}
+	if dependencies[depId].Integrity == "" {
+		dependencies[depId].Integrity = dep.Integrity
 	}
-	return false
+	dependencies[depId].Scopes = appendScopes(dependencies[depId].Scopes, scopes)
+	dependencies[depId].RequestedBy = append(dependencies[depId].RequestedBy, pathToRoot)
+	return
 }
 
-func runList(typeRestriction, executablePath, srcPath string, npmArgs []string, log utils.Log) (stdResult, errResult []byte, err error) {
-	log.Debug("Running npm list command.")
-	cmdArgs := []string{"list"}
-	cmdArgs = append(cmdArgs, npmArgs...)
+// Lookup for a dependency's tarball in npm cache, and calculate checksum.
+func calculateChecksum(cacache *cacache, name, version, integrity string, log utils.Log) (md5 string, sha1 string, sha256 string, err error) {
+	if integrity == "" {
+		var info *cacacheInfo
+		info, err = cacache.GetInfo(name + "@" + version)
+		if err != nil {
+			return
+		}
+		integrity = info.Integrity
+	}
+	var path string
+	path, err = cacache.GetTarball(integrity)
+	if err != nil {
+		return
+	}
+	return utils.GetFileChecksums(path)
+}
 
-	// These arguments must be added at the end of the command, to override their other values (if existed in nm.npmArgs)
-	cmdArgs = append(cmdArgs, "--json=true", "--all", "--"+typeRestriction)
+// Merge two scopes and remove duplicates.
+func appendScopes(oldScopes []string, newScopes []string) []string {
+	set := utils.NewStringSet(oldScopes...)
+	set.AddAll(newScopes...)
+	return set.ToSlice()
```

Comment:
I don't think the new StringSet class is necessary. Alternatively, you can change the code to the following:
```suggestion
	contained := make(map[string]bool)
	allScopes := []string{}
	for _, scope := range append(oldScopes, newScopes...) {
		if !contained[scope] {
			allScopes = append(allScopes, scope)
		}
		contained[scope] = true
	}
	return allScopes
```